### PR TITLE
rsx: Reimplement large portions of texture reconstruction (3D and Cubemap)

### LINKED ISSUE #17453: Deep Investigation Results: Black Reflections Bug in Gran Turismo 6 v1.22 BCUS99247
Hello Developers,
I am presenting a detailed analysis and debugging method that may lead to a definitive solution to the black reflection bug in GT6, caused by enabling the "Write Color Buffers" (WCB) and "Read Color Buffers" (RCB) options.

The Perfect Test Scenario: The Prize Car Scene

I discovered an exception scenario that is crucial for debugging this issue: the prize car scene.
With WCB and RCB enabled, the black reflection bug occurs in this scene, as it does everywhere else in the game.
However—and this is the crucial point—this is the only scene that can be rendered correctly with WCB and RCB disabled without resulting in a permanent black screen. The simplicity of this scene (white background, just the car model) makes it an ideal testing environment, as it generates cleaner, easier-to-analyze rendering logs compared to a garage or track full of objects. Proposed Debugging Method (A/B Testing)
I propose a direct comparison method to isolate the flaw in the WCB/RCB code:

Execution A (Buggy): Start the game and proceed to the prize car scene with WCB and RCB enabled. Do not advance beyond the scene. Save the RPCS3 log up to this point and close the game window.

Execution B (No Bugs): Now, disable WCB and RCB in the emulator. Restart the game, navigate through the menu to the last completed championship—this will skip the races (as they have already been completed) and load the prize car scene directly. This time, the car will appear perfectly, without the black reflections. Save the RPCS3 log again.

Comparative Analysis: Compare the log from "Run A" with the log from "Run B." The difference between the logs, specifically in the rendering and texture processing calls, will identify the exact process in the WCB/RCB implementation that is corrupting the reflection cube map.

Watch this video where I demonstrate both examples: https://www.youtube.com/watch?v=ONIY03N4lnk

Visual Evidence of the Root Cause

Analyzing the faulty reflection itself, it is possible to observe that the car's surface displays the entire mipmap chain of the reflection texture (cube map). The video shows several black dots, and in some of them, reflection thumbnails of varying sizes can be seen, proving that the emulator is not selecting the correct level of detail (mipmap), but rather projecting the entire texture, with all its levels, opaquely onto the car's paint. This points to a problem with how the texture's alpha channel (transparency) is being handled.
The visual evidence shows what is wrong (incorrect mipmap/alpha channel rendering). The prize car scene scenario and the log comparison method demonstrate how to find the cause in the code. The answer to fixing the black reflections throughout the game likely lies in the difference between the logs from these two runs.
I hope this detailed analysis can finally guide the developers to a permanent solution.

UPDATE - August 27, 2025: Deep Investigation Results
Following up on my initial report, I conducted an extensive, multi-hour interactive debugging session to isolate the root cause of this bug. The investigation produced definitive evidence that confirms and refines the original hypothesis.
Investigation Summary:
The core issue is 100% confirmed to be a conflict between the game's user interface (HUD) rendering and the reflection cube map when Buffers (WCB/RCB) are enabled. We verified this through a series of targeted code modifications and experiments.

1. Definitive Proof: Isolating the UI Rendering Process
The most critical experiment was to completely disable alpha blending in the Vulkan renderer by forcing attachment.blendEnable = VK_FALSE; in graphics_pipeline_state.hpp.
Result: This successfully "broke" the UI transparency, causing elements to render with opaque black backgrounds.
The smoking gun: Crucially, even with the UI rendering incorrectly (but still rendering), the black reflection bug on the car persisted. This demonstrates a direct and undeniable link: the very act of drawing the UI over the 3D scene is what causes the reflection corruption.

2. Attempted Solutions and Their Results (Evidence for Developers)
Next, we tried several surgical fixes, all of which failed, but provided critical insights into the nature of the problem:
Modifying Blending Equations: We tried forcing different blending factors (e.g., VK_BLEND_FACTOR_ZERO, VK_BLEND_FACTOR_ONE_MINUS_DST_COLOR) for the alpha channel. Each attempt simply changed the nature of the visual artifacts (e.g., causing opaque white boxes, incorrect decal transparency, or gray outlines) without fixing the underlying reflection bug. This proves that no simple change to the blending equation can solve the problem.
The Pipeline Barrier (The Final Test): Our most advanced hypothesis was that this was a synchronization issue (a race condition) where UI rendering began before the reflection buffer was finished. We successfully identified the exact location in the renderer to insert a pipeline barrier (vk::insert_global_memory_barrier in VKGSRender.cpp) just before the UI was drawn. This is the technically correct way to force the GPU to complete all 3D scene rendering before starting the UI.

Result: Even with a perfect synchronization barrier, the black reflection bug remained completely unaffected.
Final Conclusion
The failure of the pipeline barrier to solve the problem leads us to our final conclusion:
This is not a simple race condition. This is a deeper architectural issue related to how RPCS3's "Write Color Buffers" implementation handles compositing a transparent UI layer over a 3D scene that relies on the same render target. Drawing the UI with WCB/RCB enabled appears to be fundamentally altering the color buffer's alpha channel in a way that synchronization alone cannot prevent.
The problem isn't when the UI is drawn, but rather that it is being drawn under the current WCB memory management model.
I hope this thorough, step-by-step documentation of our debugging process, the elimination of variables, and the final confirmation of the bug's architectural nature will provide the official development team with the evidence they need to create a definitive solution. All the fundamental research has already been done.

UPDATE August 28, 2025  Definitive Analysis of the GT6 Black Reflection Bug: Probable Root Cause Identified in Depth-only Pass

I am presenting a final and much more precise analysis of the black reflection bug in Gran Turismo 6, caused by enabling the "Write Color Buffers" (WCB) and "Read Color Buffers" (RCB) options. The investigation using RenderDoc, which generated A/B capture files, has allowed us to isolate what is, with high probability, the root cause of the issue.

The Probable Root Cause: Corruption of the Z-Prepass
The investigation revealed that the bug is not a late-stage symptom caused by the UI, but rather a failure that occurs in one of the earliest stages of the frame's rendering pipeline: the Depth-only Pass #2.
A "Depth-only Pass" (or Z-prepass) is an optimization pass whose sole purpose is to write depth information to the Z-Buffer, without calculating or writing any color.
Normal Behavior (WCB Disabled): In the gt6_without_bug.rdc capture, the Depth-only Pass #2 behaves correctly. Its color output is non-existent or empty, and it only populates the depth buffer.
Buggy Behavior (WCB Enabled): In the gt6_with_bug.rdc capture, this same pass is fundamentally corrupted. The WCB option appears to force the pipeline state to incorrectly write colors to the framebuffer during this pass that should be "depth-only". The reflection shader is executed prematurely, before all its required resources (like the environment cubemap) are ready, resulting in the black color we see.

Conclusion
The previous hypothesis about UI corruption was not incorrect, but rather an observation of a symptom. The true root cause appears to be much deeper and earlier in the pipeline: the corruption of the Depth-only Pass #2. Enabling WCB is forcing the execution of color shaders at a stage where only depth should be written.
This discovery provides an exact and verifiable starting point for a fix. The likely solution lies in understanding why the WCB implementation is altering the rendering state (render targets and fragment shaders) during the Z-prepass stages.

I hope this focused and precise analysis can finally lead to a definitive solution.

All this research and deep debugging work was only possible with the help of the combination of Gemini 2.5 pro + Gemini 2.5 flash agents simultaneously.

UPDATE October 24, 2025

While testing a new combination of settings in RPCS3, I noticed unusual behavior: some of the black reflections simply disappeared gradually when changing the image scale:

See this video where I demonstrate how to render.

https://www.youtube.com/watch?v=pjG5L6pWeCw

UPDATE March 25, 2026

I was mistaken about the solution to the problem; changing the image scale only masks the problem but does not solve it, creating new problems such as colored artifacts and color distortion.

The black spots are empty spaces on the cube map that were not filled in time, however the black spaces are not completely empty, they contain stacked miniatures of the cube map reflection, to reproduce these effects in your emulator, you don't need to win any races in career mode or unlock anything, just start a quick race on the Willow Springs circuit on the Big Willow layout, during the day, I recommend using the courtesy Ford Mustang car, as it has a black interior and helps to better visualize the reflections, go through the first corner, pause the game and exit the race, select save replay, enter replay mode and pause it before the first corner after passing the finish line, enter photo mode and look at the reflections in the right window,
Close the game window, change the resolution scale (between 103% and 150%) in RPCS3 and save, go to the gallery, and play the saved replay, pause the replay in the same place I indicated earlier and see the spaces that would be black, now filled with Distorted textures.

I conclude that...

The alternatives I imagine to solve the problem are:

A) Implement some master trick that ensures RPCS3 can generate the image on time despite the high latency of DDR5 compared to the very low latency of the native PS3's Cell XDR + RSX GDDR3 without generating a fatal error, since in RPCS3 frame blit is occurring before the image is fully generated, resulting in an incomplete image.

B) Since the problem lies in the latency of operations, someone with a supercomputer, with a powerful CPU and very low latency RAM, combined with a powerful GPU, might be able to render the game correctly.

C) A third and more viable alternative would be to try running the game with the BC-250 motherboard, as it is a weakened PS5, and has unified GDDR6 memory. The CPU and GPU are on the same chip and share the same memory, which could hypothetically solve the black reflection problem, as there would be no delays due to DDR5 memory latency. The downside would be that the FPS could drop drastically, since the PS5 CPU (which is based on the Zen 2 architecture) does not support AVX-512 instructions, as AVX-512 instructions are important for RPCS3 emulation.

D) If everything I mentioned above is tried and doesn't work, I imagine the last alternative to solve this problem would be to use the PS3 development console called DECR 1000a to run a replay of a GT6 race and capture the complete log, then run the same replay in RPCS3 and also capture the complete log, and then compare the two logs to find a clue as to what can be done in RPCS3 to replicate the behavior of the native PS3.

UPDATE March 27, 2026
After further analysis of the causes of PCIe bus latency and the fundamental architectural differences between the PS3 and modern PCs, it is realistic to consider that, even after making all the optimizations, they may not be sufficient to solve the problem. Traditional CPU-to-GPU emulation may encounter a physical limit with the Gran Turismo 6 v1.22 game. Therefore, two additional, albeit extremely complex, "revolutionary" alternatives come to mind:

E) Emulate the entire PS3 (Cell + RSX) exclusively on the GPU. Since the main bottleneck is the communication latency between the system RAM (DDR5) and the GPU VRAM via the PCIe bus, moving the entire pipeline to the GPU could theoretically eliminate this delay. However, the caveat is enormous: while GPUs excel at parallel processing (SIMD) and can easily handle SPEs, they have significant difficulty with complex branching and conditional logic. The main Cell core (PPE), which handles the operating system and game logic, would cause a serious "warp divergence" on a modern GPU, drastically hindering performance. Implementing this would essentially require rewriting RPCS3 from scratch in a Compute Shader-based emulator, which, for example, could be the solution to graphical problems in other racing games that seem to stem from latency issues, such as some titles in the Need for Speed ​​series, like NFS The Run, NFS Rivals, NFS Hot Pursuit, NFS Most Wanted (2012).

F) Completely abandon real-time emulation in favor of static recompilation (a native PC port). Similar to what the reverse engineering community recently achieved with N64 games (such as Super Mario 64 and Zelda: Ocarina of Time). Instead of translating instructions in real time, this would involve decrypting the GT6 EBOOT.BIN, converting the PowerPC assembly code to C/C++, and translating the RSX graphics calls directly to native Vulkan. By removing the virtual machine layer, the game would run natively on x86_64 hardware, completely bypassing all the architectural bottlenecks of the PS3. The main drawback is the monumental human effort required to reverse engineer such a colossal and highly optimized codebase as GT6, although modern AI tools could theoretically be used to accelerate semantic renaming and pattern recognition of the decompiled code.

To achieve the quality and stability level of GT6 v1.22 on the PS3, Polyphony Digital had to push the game to the limits of what the hardware offered.

Emulating this game without any problems is like emulating the peak performance of the PS3 on a different architecture; it's like trying to win a race against a formidable opponent, with no second place on the podium. It's win or win; it won't be an easy task, but it's theoretically possible.

### LINKED ISSUE #18607: GT6 black reflections (NaN poisoning)
Hello everyone, after some months of digging, I've finally found the fix for the "black reflections" issue caused by enabling Read Color Buffers and Write Color Buffers in Gran Turismo 6.

Enabling these settings is required in versions after 1.05. If we don't enable them, almost the entire screen turns black during races (except HUD). One side effect of enabling them was partial corruption in reflections.

With the help from kd-11 (huge thanks!), we managed to find the issue: the reflection fragment shader path is still correct, but in GT6 reflection shaders (notably those sampling TEX14) some DIV/DIVSQ operations can end up using unstable divisors (zero / near-zero / NaN / Inf), which produces invalid math in the reflection term.

Invalid divisor values in those instructions has caused NaN/Inf propagation into reflection color computation, and this is why we see parts of the car appear black.

The fix in the code is a targeted divisor sanitization in fragment shader decompilation for GT6 reflection shaders: when the shader references TEX14 and contains DIV/DIVSQ, denominators are sanitized before division (NaN/Inf fallback, epsilon clamp for near-zero), and DIVSQ uses the same safe denominator path. This fixes the issue by preventing invalid math from collapsing reflection output to black, while keeping RCB/WCB enabled and preserving normal race rendering.

This seems to be a GT6-specific issue and fix. I also understand the emulator code should be game-agnostic, and I’m not sure how best to generalize this yet, so I will paste the 2 modified files here so we can discuss the best upstream approach.

[FragmentProgramDecompiler.zip](https://github.com/user-attachments/files/26878438/FragmentProgramDecompiler.zip)

Behavior (before this fix)

Image

behavior (after this fix)

Image

These are the changes made in FragmentProgramDecompiler.cpp:

Image

Image

Image

Image

Image

These are the changes made in FragmentProgramDecompiler.h:

Image

Quick note: all code changes are AI Generated, i was just assisting it the best i could, so of course there could be some "slop" but the thing is, this code fixes the issue. Fixes #9536 #17453  , part of #12567

### LINKED ISSUE #18757: Gran Turismo 6: HUD/cockpit mirror rendering issues – experimental workaround + public test build
### Quick summary

Narrow follow-up to #12567.

This issue tracks a focused experimental workaround for two specific Gran Turismo 6 mirror rendering problems:

* default 2D HUD rear-view mirror showing wrong content, e.g. minimap-like texture data
* car/material flicker in the default HUD mirror and cockpit mirrors

I can close this if maintainers prefer keeping it inside the existing umbrella issue.

### Test target

```text
Game: Gran Turismo 6
Title ID: BCES01893
Game version: 1.05
Base RPCS3 commit: b533a560e6df6f8aa1c271bfffbd1e39f500bd65
Embedded version before patch: 19374-b533a560
```

### Public test artifact

Repo with patch, notes, hashes and comparison video:

https://github.com/SputnikKaputtnik/rpcs3-gt6-mirror-fix-publictest

This is **not an official RPCS3 build**, **not an upstream-ready PR**, and **not intended for general support**. It is a focused public test artifact.

### What the workaround attempts

For the default HUD mirror final image:

* GT6 appears to sample a mirror-sized color target at `0xc2200000`
* in affected frames, the surface-cache overlap winner can become a later depth target instead of the exact color surface
* the patch prefers the exact color overlap for the observed `0xc2200000`, `475x98`, `pitch=8192` mirror read
* it also forces a copy for this mirror color read so the shader does not sample a still-bound render target directly

For mirror material flicker:

* GT6 mirror material passes appear to use a framebuffer-backed reflective cubemap at `0xc40c0000`
* they also use a framebuffer-backed `1x1` lighting/material scalar at `0xc0000000`
* keeping those temporary cache entries alive stabilizes resource identity across mirror draws
* the `1x1` scalar is changed from `copy_image_static` to `copy_image_dynamic` so cached hits refresh the current value instead of freezing stale lighting/material data
* cockpit mirror coverage includes observed targets/scissors around `0xc3500000`, `0xc1320000`, `0xc0660000`, and `0xc0cc0000`

### Recommended test config

```text
Read Color Buffers: true
Write Color Buffers: true
Asynchronous Texture Streaming: true
Multithreaded RSX: true
Resolution Scale: 200
```

During local A/B testing, unrelated full-screen flicker was traced to configuration and disappeared with the settings above.

### Local result

```text
Default HUD mirror: fixed in first local test
Cockpit mirrors: fixed in first local test
Unrelated full-screen flicker: gone with the config above
```

### Suggested public test coverage

* multiple GT6 tracks, especially Grand Valley East
* default 2D HUD view with rear-view mirror
* cockpit center and side mirrors
* outdoor/tunnel lighting transitions
* cold boot and savestate resume

### Cross-checks

Quick local cross-checks:

* Gran Turismo 5 Prologue: rear-view mirror path appears different; GT6 mirror bugs not observed; no visible regressions seen
* Need for Speed Shift: rear-view mirror path appears different; GT6 mirror bugs not observed; no visible regressions seen

### Known limits

* tested locally only on `BCES01893` so far
* not title-gated
* relies on observed GT6 RSX addresses, dimensions and scissor patterns
* should be treated as an experimental workaround for public testing/review, not as a merge proposal

### AI transparency

This experiment was developed with substantial assistance from OpenAI Codex/ChatGPT. Human work included bug reporting, local tests, patched/unpatched comparisons, cross-checks and publication review.

The AI-assisted code should be reviewed like any other experimental patch, with special attention to correctness, blast radius and title-specific assumptions.

### PR DESCRIPTION
**Pull request by kd-11**
Another huge set of bugs uncovered when debugging a hackfix for LBP2 (@Yahfz). Turns out so much of the texture processing logic was broken. The problems were confirmed using hardware tests and have been fixed accordingly.
Fully supports reconstructing 3D textures with mipchains from host memory (VRAM). This should fix problems that may be observed when WCB is disabled in some games, but tbh the blast radius is enormous, so test any games you may have access to.

Tests:

1. Generic 3D mipchain
[texture3d_mips.zip](https://github.com/user-attachments/files/31089145/texture3d_mips.zip)

2. Host-only 3D mipchain
[texture3d_unwrap.zip](https://github.com/user-attachments/files/31089148/texture3d_unwrap.zip)

<details>
<summary>Test results</summary>

Result (PS3):

image

Result (Master, no WCB, host-only - beware strobing)

image

Result (PR, no WCB)

image

NOTE: The strobing is expected, because only 1 frame was rendered. It is not a bug. I can fix it by drawing to all the frame outputs but the whole point of the hw tests is to trace a single operation one time and that is a lot more convenient this way.

</details>

Closes https://github.com/RPCS3/rpcs3/issues/18607
Closes https://github.com/RPCS3/rpcs3/issues/18757
Closes https://github.com/RPCS3/rpcs3/issues/17453

### CODE REVIEW

**Comment by woj1993** [comment]
What about cases that WCB is on?

**Comment by kd-11** [comment]
> What about cases that WCB is on?

Better performance I suppose since it needs to reupload fewer things. But even then, WCB doesn't catch all edge cases so there could still be improved rendering even with it enabled. WCB part is only specific to my test case and should not be assumed to stand in for all games.

**Comment by Jonathan44062** [comment]
LBP3 crashes when you try to play the levels during loading screen, master works fine 
[RPCS3.log](https://github.com/user-attachments/files/31090497/RPCS3.log)

**Comment by digant73** [comment]
as reported in PR's description, it seems some games don't need WCB (e.g. WCB not needed in Sonic Generation)

File: rpcs3/Emu/RSX/Common/TextureUtils.h [hunk 1]
```diff
@@ -345,10 +351,12 @@ namespace rsx
...
 	/**
-	* Get number of bytes occupied by texture in RSX mem
+	* Get number of bytes occupied by texture in RSX mem.
+	* Specify mip_level to compute size for exactly one mipmap level.
 	*/
-	usz get_texture_size(const rsx::fragment_texture &texture);
-	usz get_texture_size(const rsx::vertex_texture &texture);
+	constexpr u8 RSX_GCM_MIP_LEVEL_IGNORED = UINT8_MAX;
+	usz get_texture_size(const rsx::fragment_texture &texture, u8 mip_level = RSX_GCM_MIP_LEVEL_IGNORED);
+	usz get_texture_size(const rsx::vertex_texture &texture, u8 mip_level = RSX_GCM_MIP_LEVEL_IGNORED);
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Common/TextureUtils.h:359]:
```suggestion
	usz get_texture_size(const rsx::fragment_texture& texture, u8 mip_level = RSX_GCM_MIP_LEVEL_IGNORED);
	usz get_texture_size(const rsx::vertex_texture& texture, u8 mip_level = RSX_GCM_MIP_LEVEL_IGNORED);
```

**Comment by papagoghub** [comment]
GT6

This version fixes the issue of black blocks appearing on the body of the Gran Turismo 6 + Read Color Buffers.

**Comment by digant73** [comment]
I confirm the fix in GT6. Reflection on cars is now perfect on both menu and ingame.

### Master:
image

### PR:
image

**Comment by digant73** [comment]
PR seems reducing (not eliminating) some graphical issues in some areas in Resistance 2 like in chapter `Chicago, Illinois` section `Michigan Avenue`

### Master:
image

### PR:
image

**Comment by AniLeo** [comment]
We need before and after to know what was wrong before

**Comment by digant73** [comment]
> We need before and after to know what was wrong before

added screenshots. That issue seems a lot reduced in the PR

**Comment by kd-11** [comment]
Paused this. The crashes are caused by some insane tech debt that I failed to clean up for years. I'll fix that first on master before merging this.

**Comment by kd-11** [comment]
https://github.com/RPCS3/rpcs3/issues/19232 tracks the refactor that needs to land in master before this work.

**Comment by woj1993** [comment]
So now you need to rebase this on master and give us for testing?

**Comment by kd-11** [comment]
> So now you need to rebase this on master and give us for testing?

Rebased and the crashes fixed.

**Comment by kd-11** [comment]
> this PR or #19235 reduced GPU usage (reported by RTSS) on GT6 (from 40% to %30). Testing other games...

Probably this PR. It allows to grab things from VRAM instead of doing round-trip to CPU each time.
I have found a couple of crashes (verification) that I am currently fixing though. Fixes pushed, hopefully no more asserts.

**Comment by digant73** [comment]
> > this PR or #19235 reduced GPU usage (reported by RTSS) on GT6 (from 40% to %30). Testing other games...
> 
> Probably this PR. It allows to grab things from VRAM instead of doing round-trip to CPU each time. I have found a couple of crashes (verification) that I am currently fixing though. Fixes pushed, hopefully no more asserts.

I had WCB enabled on pre #19235 so it increased a little bit GPU usage. however even without WCB, GPU seems a little bit better on one of the mentioned PRs

**Comment by digant73** [comment]
closes #18757 (fixed in the past by #19090)
closes #18607
closes #17453

**Comment by digant73** [comment]
I did some no-regression tests and I didn't see any graphical or performance regression, but I would wait feedbacks from other users

**Comment by TigiHof** [comment]
Did some tests as well (Uncharted 2, FEAR 2, GoW 3, Eternal Sonata, Lollipop Chainsaw, GoT) - no regressions found, either.
